### ISSUE TITLE
Question About Speed

### ISSUE THREAD

**Issue by Vangelis27** [ThePBone/CrunchyrollDownloader#14]
When downloading an episode, ffmpeg doesn't use my max download speed provided by my isp. My download speed is 46Mbps but when i have task manager opened i see a lot of spikes in speed. Also the device i am using is the only one in then network.

![Annotation 2020-04-17 162253](https://user-images.githubusercontent.com/11654086/79574491-e579ca00-80c8-11ea-92ef-da238de2583d.jpg)

**Comment by ThePBone** [comment]
Same here, my actual download speed is 150mbps but FFmpeg still downloads at ~50mbps. I'm pretty sure that Crunchyroll's servers may have a limited download speed per IP/user. Under normal circumstances, for example, the video player in your browser buffers an episode bit by bit and does not download the whole thing at once.
I don't know if that is related to those spikes though, but it's very likely that CR has this limit in place.

**Comment by Vangelis27** [comment]
Is it possible that the speed is bottlenecked by our cpu? Cpu usage is not that high that's why i suspect that it a single core performance issue. I noticed that when i open other programs that don't use my internet speed, my speed drops from ~2.47x to ~2.28x. Also Bypass geoblock seems to be working fine today.

**Comment by ThePBone** [comment]
I just remembered why there are these peaks in the task manager.

M3U8 HLS streaming playlists have their actual media file split in several MPEG/TS files (about ~150 parts for one 24-minute episode). So after FFmpeg finished one of those parts it needs to analyze the next one, before it can start to download and to append it to the previous parts. Because of that FFmpeg will pause the encoding process until the parts were fully analyzed. That would explain those spikes in the task manager.

> Is it possible that the speed is bottlenecked by our cpu?

I'm not sure. Running FFmpeg with `-threads <count>` did not change anything. 

Also, there isn't even much encoding going on so there shouldn't be a high CPU usage. The only thing that is being actually processed/encoded by FFmpeg is the audio (I'm using an audio bitstream filter to convert the MPEG-2 ADTS to MPEG-4 headers). The video parts are directly copied and appended.

**Comment by Vangelis27** [comment]
I see. So there's no way of speeding up the process?

**Comment by ThePBone** [comment]
Unfortunately, no. It doesn't seem that way.

**Comment by Vangelis27** [comment]
Is it possible to port the program to an android phone ? I know it's irrelevant to the topic but i didn't want to clutter the issues list.

**Comment by ThePBone** [comment]
Yeah, it is possible. But I don't know how well FFmpeg on android works, it might have to be replaced with something else.

**Comment by Vangelis27** [comment]
I have an idea, since it can use our max download speed, is it possible to download multiple episodes at the same time?

**Comment by ThePBone** [comment]
I just tried to download three episodes at once (using three FFmpeg instances) and for each episode FFmpeg displayed a download speed of >5.0x. So yes, your idea would work pretty well.
I think it would be a cool feature to add parallel downloads but I'd need to rewrite most parts of the code that handle the FFmpeg conversion job. Therefore, I don't know when I'll add it and how long it will take to implement it (contributions are welcome ofc).

But for now, I recommend you to use multiple instances of the downloader at the same time to finish downloads faster.

By the way, I've begun to port the downloader to Android. Again, I don't have an ETA because I don't work on it every day but the base UI is finished and it is already capable of directly streaming episodes in VLC, MX Player, etc...

**Comment by Vangelis27** [comment]
Wouldn't it be better if you made the multiple downloads first so that you can port it to android ?

**Comment by ThePBone** [comment]
No, unfortunately I can't port the Desktop version directly to Android. Apart from the programming language differences, the mobile app has a completely new structure (optimized for small screens) and needs to be written from scratch so I can't reuse old code snippets.

But I'll probably add the multiple download feature to the desktop version first anyway.
